File: 10.validParenthesis.py
#given a string containing only parenthesis,
#determine if it is valid (all closed). 

def validP(givenString):
    leftSet = set('([{')
    rightSet = set(')]}')
    
    match = { 
             ")" : "(",
             "]" : "[",
             "}" : "{" 
            }
    
    stack = []

    for bracket in givenString:
        if bracket in leftSet:
            stack.append(bracket)
        if bracket in rightSet:
            if not stack:
                return False
            elif stack.pop() != match[bracket]:
                return False
            else:
                continue
    if not stack:
        return True
    else:
        return False
  



print(validP("({[(]})"))
print(validP("())"))
print(validP("()[{}]()"))
print(validP("()"))

File: 10.validParenthesis.js
//given a string containing only parenthesis,
//determine if it is valid (all closed). 

const validP = (givenString) => {
    const match = {
        "(" : ")",
        "[" : "]",
        "{" : "}"
    }

    if(givenString.length === 0) return false

    const stack = []

    for(let i = 0; i< givenString.length; i++){
        if(match[givenString[i]]){
            stack.push(givenString[i])
        } else {
            const leftBracket = stack.pop()
            const correctBracket = match[leftBracket]

            if(givenString[i] != correctBracket){
                return false
            }
        }
    }
    return stack.length === 0
}


console.log(validP("({[(]})"))
console.log(validP("())"))
console.log(validP("()[{}]()"))
console.log(validP("()"))


